text/3753-rest-pattern-matches-non-struct.md

```diff
@@ -0,0 +1,138 @@
+- Feature Name: `rest_pattern_matches_non_struct`
+- Start Date: 2024-12-31
+- RFC PR: [rust-lang/rfcs#3753](https://github.com/rust-lang/rfcs/pull/3753)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Special case struct patterns containing only a rest pattern (e.g., `Foo { .. }`)
+so that they can match values of any type with the appropriate name, not just
+structs (e.g., it could match an `enum Foo` value). This is done so that structs
+containing only private fields can be changed to other types without breaking
+backwards compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+It is common for a library's API to have a public struct with all-private
+fields, with the intention of making the type opaque, hiding its contents as
+implementation details. For example:
+```rust
+// A library's initial API
+pub struct Foo {
+    inner: FooInner,
+}
+enum FooInner {
+    A,
+    B,
+}
+```
+
+In a later version, the library might want to expose the internals of the API.
+For example:
+```rust
+// The library's API is later changed to this.
+pub enum Foo {
+    A,
+    B,
+}
+```
+
+Intuitively, one might think that this change is backwards-compatible. However,
+this is technically not the case, since client code might match the `Foo` type
+with a rest pattern (`Foo { .. }`), which currently only matches structs, and
+not enums or other types.
+```rust
+// Client code that uses the library
+// Works with the initial API, but doesn't work with the later API.
+fn do_something(x: Foo) {
+    match x {
+        Foo { .. } => {}
+    }
+}
+```
+
+To eliminate this semver hazard, this RFC proposes that the pattern `Foo { .. }`
+should match values of any type named `Foo`, not just structs.
```

Comment:
This should also mention motivation for macros generating match statements where they want to verify a user-passed type.

Reply:
Can't a match arm like this already be generated?
```rust
ref foo_val => {
    let _: &Foo = foo_val;
}
```